### PR TITLE
feat: [FC-7879] add signal handler to save assignment dates to Conten…

### PR DESCRIPTION
**Pull request by kyrylo-kh**
> NOTE: Depends on #343

## Description
Adds a new signal handler to support saving assignment due dates to the `ContentDate` model in the `edx-when` service.

### Key Changes:
* Introduces the `_Assignment` data class to standardize assignment metadata (title, date, block key, type).
* Implements `update_or_create_assignments_due_dates()` to:
  
  * Create or update `ContentDate` entries with assignment titles, dates, and metadata.
  * Store related information like `block_type`, `subsection_name`, and `course_name`.
  * Ensure `DatePolicy` entries are reused or created as needed.
* Adds logging and input validation to handle incomplete or invalid assignment data gracefully.

### Testing:
Includes a full test suite verifying:

* New entries are inserted correctly.
* Existing entries are updated when data changes.
* Graceful handling of:
  
  * Assignments with missing title or date.
  * Multiple assignments at once.
  * Failures during `DatePolicy` creation.

### CODE REVIEW

**Comment by e0d** [comment]
@kyrylo-kh sorry, there also seem to be a bunch of failing tests.  Perhaps a rebase will also fix this, but please have a look.

**Review by e0d — CHANGES_REQUESTED** [review]

File: edx_when/api.py [hunk 1]
```diff
@@ -543,6 +507,56 @@ def get_schedules_with_due_date(course_id, assignment_date):
...
+            )
+            continue
+        models.ContentDate.objects.update_or_create(
+            course_id=course_key,
+            location=assignment.block_key,
+            field='due',
+            block_type=assignment.assignment_type,
+            defaults={
+                'policy': models.DatePolicy.objects.get_or_create(abs_date=assignment.date)[0],
+                'assignment_title': assignment.title,
```
[the remark this answers is not in the document]
Comment (kyrylo-kh) [reply edx_when/api.py]: Do you mean it can confuse having subsection_name in model for different type of content?
We wrote description of this in argument:
```
        assignments: List of Assignment instances. Assignments with missing date or
            title are skipped (with a warning). Use Assignment.subsection_name for
            the containing subsection when available; it is stored separately from
            assignment_title.
```

File: edx_when/api.py [hunk 2]
```diff
@@ -548,6 +549,77 @@ def get_schedules_with_due_date(course_id, assignment_date):
...
+                continue
+            models.ContentDate.objects.update_or_create(
+                course_id=course_key,
+                location=assignment.block_key,
+                field='due',
+                block_type=assignment.assignment_type,
+                defaults={
+                    'policy': models.DatePolicy.objects.get_or_create(abs_date=assignment.date)[0],
+                    'assignment_title': assignment.title,
+                    'course_name': course_key.course,
```
Comment (edschema) [inline edx_when/api.py]: Docstring says course_key can be a string; if a string is passed to this fn, a string caller will raise an attribute error. Suggestion: normalize `course_key` with " _ensure_key(CourseKey, course_key)"

File: edx_when/api.py [hunk 3]
```diff
@@ -548,6 +549,77 @@ def get_schedules_with_due_date(course_id, assignment_date):
...
+                    "Skipping assignment '%s' for course %s because it has no date or title",
+                    assignment,
+                    course_key_str
+                )
+                continue
+            models.ContentDate.objects.update_or_create(
+                course_id=course_key,
+                location=assignment.block_key,
+                field='due',
+                block_type=assignment.assignment_type,
```
Comment (edschema) [inline edx_when/api.py]: `block_type` is used elsewhere as the structural XBlock type. Suggestion: define `assignment_type` if we need. But a more general question: Is the block_type or assignment_type needed to identify an existing date row?

Comment (kyrylo-kh) [reply edx_when/api.py]: Identity question: No. Row identity = (location, field) (unique_together policy,location,field; location encodes course+block). block_type/assignment_type not needed to match.
I moved block_type out of lookup to defaults. https://github.com/openedx/edx-when/pull/347/changes/c875945f14100f5634a8d1d5f9c237cc2f59b74b
Value now assignment.block_key.block_type (structural: 'sequential'), matching usage everywhere else

File: edx_when/api.py (hunk 2 above)
Comment (kyrylo-kh) [reply edx_when/api.py]: Fixed - https://github.com/openedx/edx-when/pull/347/changes/c875945f14100f5634a8d1d5f9c237cc2f59b74b

File: edx_when/api.py [hunk 4]
```diff
@@ -548,6 +549,78 @@ def get_schedules_with_due_date(course_id, assignment_date):
...
+                    assignment,
+                    course_key_str
+                )
+                continue
+            models.ContentDate.objects.update_or_create(
+                course_id=course_key,
+                location=assignment.block_key,
+                field='due',
+                defaults={
+                    'policy': models.DatePolicy.objects.get_or_create(abs_date=assignment.date)[0],
```
Comment (pkulkark) [inline edx_when/api.py]: Won't this reintroduce the race condition noted in line 430?

File: edx_when/api.py [hunk 5]
```diff
@@ -548,6 +549,78 @@ def get_schedules_with_due_date(course_id, assignment_date):
...
+@dataclass
+class Assignment:
+    """
+    Represents an assignment with a title, date, block key, assignment type, and optional subsection name.
+    """
+
+    title: str
+    date: datetime | None
+    block_key: UsageKey
+    assignment_type: str
```
Comment (pkulkark) [inline edx_when/api.py]: This field doesn't seem to be persisted. Instead `assignment.block_key.block_type` is used.  Either persist it or remove it from the dataclass.

File: edx_when/api.py [hunk 6]
```diff
@@ -548,6 +549,78 @@ def get_schedules_with_due_date(course_id, assignment_date):
...
+            title are skipped (with a warning). Use Assignment.subsection_name for
+            the containing subsection when available; it is stored separately from
+            assignment_title.
+
+    Returns:
+        None
+    """
+    course_key = _ensure_key(CourseKey, course_key)
+    course_key_str = str(course_key)
+    with transaction.atomic():
```
Comment (GlugovGrGlib) [inline edx_when/api.py:603]: Will be good to have a test case for the atomic boundary in this method

File: edx_when/api.py [hunk 7]
```diff
@@ -548,6 +549,78 @@ def get_schedules_with_due_date(course_id, assignment_date):
     return schedules
 
 
+@dataclass
+class Assignment:
```
Comment (GlugovGrGlib) [inline edx_when/api.py:558]: suggestion (optional): this dataclass can be extracted into data.py

File: edx_when/api.py (hunk 4 above)
Comment (GlugovGrGlib) [reply edx_when/api.py]: Agree, as @pkulkark mentioned, this get_or_create for policy may be problematic due to race condition with `DatePolicy` creation mentioned in [set_date_for_block](https://github.com/openedx/edx-when/blob/acb1a0cc83d9e802719bfdf49a15002e3ac0c855/edx_when/api.py#L428).

I believe it's necessary to reuse [_set_content_date_policy](https://github.com/openedx/edx-when/blob/acb1a0cc83d9e802719bfdf49a15002e3ac0c855/edx_when/api.py#L428) from `set_date_for_block`.
Also, it will be good to add a test case against DatePolicy duplicate.

File: tests/test_api.py [hunk 8]
```diff
@@ -747,3 +749,162 @@ def test_relative_dates_disabled(self):
     @patch.dict(sys.modules, {'openedx.features.course_experience': None})
     def test_relative_dates_import_error(self):
         assert not api._are_relative_dates_enabled()  # pylint: disable=protected-access
+
+
+class TestUpdateAssignmentDatesForCourse(TestCase):
+    """
+    Tests for the update_assignment_dates_for_course task.
```
Comment (GlugovGrGlib) [inline tests/test_api.py]: Looks like the classname and docstring are not correct as this test cases are for the `update_or_create_assignments_due_dates`

File: tests/test_api.py [hunk 9]
```diff
@@ -747,3 +749,162 @@ def test_relative_dates_disabled(self):
...
+            course_id=self.course_key,
+            location=self.block_key
+        )
+        self.assertEqual(content_date.assignment_title, 'Updated Assignment')
+        self.assertEqual(content_date.subsection_name, 'Updated Subsection')
+        self.assertEqual(content_date.policy.abs_date, self.due_date)
+
+    def test_assignment_with_null_date(self):
+        """
+        Test handling assignments with null dates.
```
Comment (GlugovGrGlib) [inline tests/test_api.py]: Suggest adding a case when row already exists and null date is passed

File: tests/test_api.py [hunk 10]
```diff
@@ -747,3 +749,162 @@ def test_relative_dates_disabled(self):
...
+
+
+class TestUpdateAssignmentDatesForCourse(TestCase):
+    """
+    Tests for the update_assignment_dates_for_course task.
+    """
+
+    def setUp(self):
+        self.course_key = CourseKey.from_string('course-v1:edX+DemoX+Demo_Course')
+        self.course_key_str = str(self.course_key)
```
Comment (GlugovGrGlib) [inline tests/test_api.py]: `self.course_key_str` and `self.staff_user` are not used in the test cases, consider removing them

**Review by GlugovGrGlib — CHANGES_REQUESTED** [review]
Thanks, left several suggestions.

In addition, if this PR is planned to be merged after the https://github.com/openedx/edx-when/pull/346, it's necessary to include version bump and a changelog update for both changes.

**Review by GlugovGrGlib — APPROVED** [review]
Code looks good, there are two points that should be done before merging it:
- Improve coverage for the patch to the target values
- Update Changelog and bump version after rebase on top of #346

**Comment by farhaanbukhsh** [comment]
@kyrylo-kh  can you rebase and run the tests again :)

**Comment by kyrylo-kh** [comment]
@farhaanbukhsh please, check again

**Comment by farhaanbukhsh** [comment]
@kyrylo-kh please wait for https://github.com/openedx/edx-when/pull/371 to land in and then if you rebase we can fix the CI issues :) thank you so much!

**Comment by kyrylo-kh** [comment]
@farhaanbukhsh Done

**Review by farhaanbukhsh — APPROVED** [review]
👍

✅ I tested this PR.
✅ I read through the code
❌  I checked for accessibility issues
❌  Includes documentation

**Comment by farhaanbukhsh** [comment]
@feanil  can you please help me merge this PR. Also, I am maintaining this package https://discuss.openedx.org/t/core-contributor-rights-expansion-farhaan-bukhsh/12732/3, but I think somehow the permission got revoked to merge.

**Comment by feanil** [comment]
@farhaanbukhsh it looks like the loop wasn't fully closed on that, can you update [the catalog-info.yaml](https://github.com/openedx/edx-when/blob/master/catalog-info.yaml#L10) and file an issue in https://github.com/openedx/axim-engineering/issues/new/choose to update your access?

**Comment by feanil** [comment]
I'll merge and release this based on the other reviews.

**Review by feanil — APPROVED** [review]
